### PR TITLE
Endpoint to update supporter plus amount

### PR DESCRIPTION
**Pull request by jonflynng**
An endpoint to update the Supporter Plus amount, for both V1 and V2.

Tested on CODE on V1 and V2 subs.

### CODE REVIEW

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        charge.productRatePlanChargeId == ids.annualV2.contributionProductRatePlanChargeId.value,
+    )
+    supporterPlusCharge
+      .map(ZIO.succeed(_))
+      .getOrElse(
+        ZIO.fail(InternalServerError("Supporter Plus rate plan charge not found on subscription")),
+      )
+  }
+
+  def asSingle[A](list: List[A], message: String): IO[ErrorResponse, A] =
```
Comment (jonflynng) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala:174]: could move these into a utils file as they're used elsewhere?

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    OutputBody,
+    Any,
+    ZIOApiGatewayRequestHandler.TIO,
+  ] = {
+    val subscriptionNameCapture: EndpointInput.PathCapture[String] = {
+      EndpointInput.PathCapture[String](
+        Some("subscriptionName"),
+        implicitly,
+        EndpointIO.Info.empty.copy(
+          description = Some("Name of subscription to be moved to a different product."),
```
Comment (johnduffell) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala:99]:
```suggestion
          description = Some("Name of supporter plus subscription to have its contribution amount updated."),
```

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+      Unit,
+      OutputBody,
+      Any,
+    ] =
+      endpoint.post
+        .in("update-supporter-plus-amount")
+        .in(subscriptionNameCapture)
+        .in(
+          jsonBody[ExpectedInput].copy(info =
+            EndpointIO.Info.empty[ExpectedInput].copy(description = Some("Definition of required movement.")),
```
Comment (johnduffell) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala:116]:
```suggestion
            EndpointIO.Info.empty[ExpectedInput].copy(description = Some("Definition of the updated total amount of regular payment")),
```

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  def getNewRatePlanAmount(
+      charges: NonEmptyList[RatePlanCharge],
+      ids: SupporterPlusZuoraIds,
+      postData: ExpectedInput,
+  ): BigDecimal =
+    charges.exists(charge =>
+      charge.productRatePlanChargeId == ids.monthlyV2.contributionProductRatePlanChargeId.value ||
+        charge.productRatePlanChargeId == ids.annualV2.contributionProductRatePlanChargeId.value,
+    ) match {
+      case true => postData.newPaymentAmount - 10
```
Comment (johnduffell) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala:200]: might cause trouble if/when price rise happens

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+  private[productmove] def subscriptionUpdateAmount(subscriptionName: SubscriptionName, postData: ExpectedInput): ZIO[
+    GetSubscription with GetAccount with SubscriptionUpdate with SQS with Stage,
+    ErrorResponse,
+    OutputBody,
+  ] = {
+    (for {
+      _ <- ZIO.log(s"PostData: ${postData.toString}")
+      stage <- ZIO.service[Stage]
```
Comment (johnduffell) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala:211]:
```suggestion
      stage <- ZIO.service[Stage].map(stage => config.Stage(stage.toString))
```

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+      subscription <- GetSubscription.get(subscriptionName)
+      accountFuture <- GetAccount.get(subscription.accountNumber).fork
+
+      _ <- ZIO.log(s"Subscription is $subscription")
+
+      zuoraIds <- ZIO
+        .fromEither(
+          ZuoraIds.zuoraIdsForStage(config.Stage(stage.toString)).left.map(InternalServerError(_)),
+        )
```
Comment (johnduffell) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/updateamount/UpdateSupporterPlusAmountEndpoint.scala:222]:
```suggestion
      zuoraIds <- ZIO.fromEither(
        ZuoraIds.zuoraIdsForStage(stage).left.map(InternalServerError(_)),
      )
```

File: handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/GetSubscription.scala [hunk 7]
```diff
@@ -42,6 +42,7 @@ object GetSubscription {
       productName: String,
       ratePlanName: String,
       ratePlanCharges: List[RatePlanCharge],
+      lastChangeType: Option[String],
```
Comment (johnduffell) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/GetSubscription.scala:45]: just a note (to myself as much as others) to remember we don't need to have one `GetSubscription` to cover all the use cases- we can have multiple simpler ones if it avoids complexity overall

**Comment by johnduffell** [comment]
looks good and quite easy to understand, thanks Jon - I guess this PR is for me to check out and get merged once I have my laptop so I will get it checkout tomorrow once I have my new laptop!

**Review by michaelbjacobson — APPROVED** [review]

**Pull request merged**